Logging improvement

RedBearAK opened the pull request:
Show all newly available properties in logging output, and reformat display of active keymaps. 

Sample output: 

```
(DD) WM_CLS = 'Gnome-terminal' | WM_NME = './bin/keyszer --flush -w -v -c ~/.config/kinto/kinto.py'
(DD) DVN = 'AT Translated Set 2 keyboard' | CLK = 'False' | NLK = 'False'
(DD) Active Keymaps = 
(DD)  KMAP: [OptSpecialChars toggles]
(DD)  KMAP: [OptSpecialChars - US]
(DD)  KMAP: [User hardware-specific customizations]
(DD)  KMAP: [Wordwise - not vscode]
(DD)  KMAP: [General Terminals]
(DD)  KMAP: [General GUI]
(DD)   COMBO: RCtrl-A => Ctrl-Shift-A [General Terminals]
```

<!--- Provide a quick summary of your changes in the Title above -->

### Changes

<!-- Reference a related issue (if one exists) so things are linked nicely: -->

<!-- `Resolves #12345`, etc. Then describe your changes... -->

**Checklist**

- [ ] Added tests (if necessary)
- [ ] Updated docs or README...

joshgoebel commented:
I like the idea here, but this is too verbose... if we have a "heading" we don't need to repeat `KMAP:` over and over, and if we have a single keymap (or two) this is still super verbose... better would be to dynamically output names of maps and split them up based on their lenght (assuming a width of 80 or something):

```
(DD) WM_CLS = 'Gnome-terminal' | WM_NME = './bin/keyszer --flush -w -v -c ~/.config/kinto/kinto.py'
(DD) DVN = 'AT Translated Set 2 keyboard' | CLK = 'False' | NLK = 'False'
(DD) Active Keymaps =  ["OptSpecialChars toggles", "OptSpecialChars - US", 
(DD)      "User hardware-specific customizations", "Wordwise - not vscode", "General Terminals", 
(DD)      "General GUI"]
(DD) COMBO: RCtrl-A => Ctrl-Shift-A [General Terminals]
```

I'm not even sure "active" matters that much (for debugging) if only one of the maps actually has the key in question...  I need to look at how that is calculated again... are active maps just any map whose conditional passed?

RedBearAK commented:
> are active maps just any map whose conditional passed?

I assume that's how it works. And it does show which keymap the combo came from on the COMBO line. 

I'll play around with making the list of active keymaps more compact. Although that's already a lot like the long comma-separated list that it was generating before I modified it. It's just not very readable that way.

RedBearAK commented:
@joshgoebel 

Would this formatting be good enough? It's not calculating line length but just using a line break for every _other_ list element, which should generally land within a reasonable line length. Combining it back with the KMAPS header limits the number of lines dedicated to the active keymaps to less than half the original example. With alignment making it clear that the additional lines are just extensions of the KMAPS line. 

```
(DD) WM_CLS = 'Gnome-terminal' | WM_NME = './bin/keyszer --flush -w -v -c ~/.config/kinto/kinto.py'
(DD) DVN = 'AT Translated Set 2 keyboard' | CLK = 'False' | NLK = 'False'
(DD) KMAPS = [OptSpecialChars toggles, OptSpecialChars - US,
(DD)          User hardware-specific customizations, Wordwise - not vscode,
(DD)          General Terminals, General GUI]
(DD) COMBO: RCtrl-A => Ctrl-Shift-A [General Terminals]
```

```
(DD) WM_CLS = 'Org.gnome.Nautilus' | WM_NME = 'Applications'
(DD) DVN = 'AT Translated Set 2 keyboard' | CLK = 'True' | NLK = 'False'
(DD) KMAPS = [OptSpecialChars toggles, OptSpecialChars - US,
(DD)          User hardware-specific customizations, Overrides for Nautilus - Finder Mods,
(DD)          General File Managers - Finder Mods, Wordwise - not vscode,
(DD)          Cmd+Dot not in terminals, General GUI]
(DD) COMBO: RCtrl-UP => Alt-UP [General File Managers - Finder Mods]
```

There's an extra `elif` to account for the possibility of an odd number of keymaps, seems to work as expected if I add another "always active" keymap to my config to force an odd number of active keymaps. So, should be good to go. 

```
(DD) WM_CLS = 'Org.gnome.Nautilus' | WM_NME = 'Home'
(DD) DVN = 'AT Translated Set 2 keyboard' | CLK = 'False' | NLK = 'False'
(DD) KMAPS = [OptSpecialChars toggles, OptSpecialChars - US,
(DD)          General testing extra keymap, User hardware-specific customizations,
(DD)          Overrides for Nautilus - Finder Mods, General File Managers - Finder Mods,
(DD)          Wordwise - not vscode, Cmd+Dot not in terminals,
(DD)          General GUI]
(DD) COMBO: RCtrl-DOT => ESC [Cmd+Dot not in terminals]
```

RedBearAK commented:
Adding another `elif` to cover the case of having only a single active keymap in the list. Tested with a nearly blank config file with a single keymap inside. Looks good to me. 

```
(DD) WM_CLS = 'Gnome-terminal' | WM_NME = './bin/keyszer -v -c ~/.config/kinto/kinto\ copy.py'
(DD) DVN = 'AT Translated Set 2 keyboard' | CLK = 'False' | NLK = 'False'
(DD) KMAPS = [testing single keymap]
(DD) COMBO: A => X [testing single keymap]
```

RedBearAK commented:
Switched to colons instead of equal signs, and added "in KMAP:" to the combo line. Looks more consistent with other output lines this way. And a bit more compact. 

```
(DD) WM_CLS: 'Code' | WM_NME: 'transform.py - keyszer (Workspace) - Visual Studio Code'
(DD) DVN: 'AT Translated Set 2 keyboard' | CLK: 'False' | NLK: 'False'
(DD) KMAPS: [OptSpecialChars toggles, OptSpecialChars - US,
(DD)          User hardware keys, Code/VSCodium/Code-OSS,
(DD)          Cmd+Dot not in terminals, General GUI]
(DD) COMBO: RCtrl-TAB => [<ComboHint.BIND: 1>, Alt-TAB] in KMAP: [General GUI]
```

```
(DD) WM_CLS: 'Gnome-terminal' | WM_NME: 'kris@fedora-3700u:~/Documents/keyszer'
(DD) DVN: 'AT Translated Set 2 keyboard' | CLK: 'False' | NLK: 'False'
(DD) KMAPS: [User hardware keys, Wordwise - not vscode,
(DD)          General Terminals, General GUI]
(DD) COMBO: RCtrl-LShift-LEFT_BRACE => Ctrl-PAGE_UP in KMAP: [General GUI]
```

RedBearAK commented:
Sample output of updated code: 

```
(DD) WM_CLS: '' | WM_NME: 'gnome-shell'
(DD) DVN: 'AT Translated Set 2 keyboard' | CLK: 'False' | NLK: 'False'
(DD) KMAPS: ['OptSpecialChars toggles', 'OptSpecialChars - US',
             'User hardware keys', 'Wordwise - not vscode',
             'Cmd+Dot not in terminals', 'General GUI overrides - not Chromebook',
             'General GUI overrides - Fedora', 'General GUI overrides - GNOME',
             'General GUI']
(DD) COMBO: RCtrl-TAB => [<ComboHint.BIND: 1>, Alt-TAB] in KMAP: ['General GUI']
(DD) spent modifiers [<Key.RIGHT_CTRL: 97>]
(DD) refusing to engage second sticky bind over existing sticky bind
(OO) press TAB 1677311831.2327335
(OO) release TAB 1677311831.2328606
```

RedBearAK commented:
I don't really like the idea of not lining up the rest of the elements under the first one, but reducing the indent can make it more compact by allowing a max of about 76 characters instead of 64. Still using one space for indenting here. 

Of course, it helps that I shortened up the length of some of my keymap names. 

```py
(DD) WM_CLS: 'Gnome-terminal' | WM_NME: 'kris@fedora-3700u:~/Documents/keyszer'
(DD) DVN: 'AT Translated Set 2 keyboard' | CLK: 'False' | NLK: 'False'
(DD) KMAPS: ['User hardware keys', 'Wordwise - not vscode',
 'GenTerms overrides: Ubuntu/Fedora', 'GenTerms overrides: GNOME',
 'General Terminals', 'GenGUI overrides: not Chromebook',
 'GenGUI overrides: Fedora', 'GenGUI overrides: GNOME', 'General GUI']
(DD) COMBO: RCtrl-LShift-RIGHT_BRACE => Ctrl-PAGE_DOWN in KMAP: ['General GUI']


(DD) WM_CLS: 'gnome-text-editor' | WM_NME: 'The output is like this (Draft) - Text Editor'
(DD) DVN: 'AT Translated Set 2 keyboard' | CLK: 'False' | NLK: 'False'
(DD) KMAPS: ['OptSpecialChars toggles', 'OptSpecialChars - US', 'User hardware keys',
 'GNOME Text Editor', 'Wordwise - not vscode', 'Cmd+Dot not in terminals',
 'GenGUI overrides: not Chromebook', 'GenGUI overrides: Fedora',
 'GenGUI overrides: GNOME', 'General GUI']
(DD) COMBO: RCtrl-TAB => [<ComboHint.BIND: 1>, Alt-TAB] in KMAP: ['General GUI']
```

It's just a tweak of one number. I haven't added this change to the PR. But I will if you like this output variation. 

```py
            # indent = " " * ( len(prefix) + 2 )
            indent = " " * ( len(prefix) - 10 )
```

joshgoebel commented on src/keyszer/transform.py:
```diff
@@ -445,13 +445,30 @@ def transform_key(key, action, ctx):
 
         if logger.VERBOSE:
```
This has grown so massive it really should be pulled out into a helper and passed whatever it needs:

```py
if logger.verbse:
  log_combo_context(combo, ctx, .,,)
```

RedBearAK commented:
> This has grown so massive it really should be pulled out into a helper and passed whatever it needs:

Reasonable, and I already broke it out into a function defined just after the diagnostics dumper. That seemed like a logical spot. Needs all this to be passed to it, as far as I can tell: 

```py
def log_combo_context(combo, ctx, keymap, _active_keymaps):
```

Works fine. 

But I was not happy with the inflexibility of the logic when I changed the indent, so I went back and wrestled with the AI for a while. It just kept giving me more and more complex solutions to fix edge cases while keeping the line length checks correct on the first line and all other lines.

I realized that the desire to avoid using another line was making the problem much more difficult than it should be. So I tried to look at it a different way, and if we let the whole list be by itself rather than shoehorning part of it into the first line, it makes the problem much less intractable. We can use the Python `textwrap` module, and this is all there is to it: 

```py
    import textwrap

    debug(f'ACTIVE KEYMAPS:')
    indent = ' ' * 5
    max_len = max(80 - len(indent), 64)
    wrapped_items = textwrap.wrap(", ".join([f"'{item}'" for item in keymap_names]), width=max_len)
    output_str = f"{indent}{wrapped_items[0]}"
    for item in wrapped_items[1:]:
        if not item.startswith("'"):
            item = ' ' + item
        output_str += f"\n{indent}{item}"
    print(output_str)
```

Changing the indent will reduce the max length per line, but now only down to a minimum of 64 characters, controlled with `max()`. One space is added if necessary to make it more clear when a list item is wrapped within itself, rather than at the item-item boundary. 

This looks like this, with a 5 spaces indent: 

```sh
(DD) WM_CLS: '' | WM_NME: 'gnome-shell'
(DD) DVN: 'AT Translated Set 2 keyboard' | CLK: 'False' | NLK: 'False'
(DD) ACTIVE KEYMAPS:
     'OptSpecialChars toggles', 'OptSpecialChars - US', 'User hardware keys',
     'Wordwise - not vscode', 'Cmd+Dot not in terminals', 'GenGUI overrides: not
      Chromebook', 'GenGUI overrides: Fedora', 'GenGUI overrides: GNOME',
     'General GUI'
(DD) COMBO: RCtrl-TAB => [<ComboHint.BIND: 1>, Alt-TAB] in KMAP: 'General GUI'


(DD) WM_CLS: 'Gnome-terminal' | WM_NME: 'kris@fedora-3700u:~/.config/toshy'
(DD) DVN: 'AT Translated Set 2 keyboard' | CLK: 'False' | NLK: 'False'
(DD) ACTIVE KEYMAPS:
     'User hardware keys', 'Wordwise - not vscode', 'GenTerms overrides:
      Ubuntu/Fedora', 'GenTerms overrides: GNOME', 'General Terminals', 'GenGUI
      overrides: not Chromebook', 'GenGUI overrides: Fedora', 'GenGUI overrides:
      GNOME', 'General GUI'
(DD) COMBO: LShift-RCtrl-LEFT_BRACE => Ctrl-PAGE_UP in KMAP: 'General GUI'
```


I'll still have all of this in the separate function, but if you're good with this style of formatting the output at least  the logging function code will be much simpler.

RedBearAK commented:
Updated PR with the simplest version of the function my eyes can come up with.

joshgoebel commented on src/keyszer/transform.py:
```diff
@@ -216,6 +216,31 @@ def dump_diagnostics():
+    max_len = max(80 - len(indent), 64)
+    wrapped_items = textwrap.wrap(", ".join([f"'{item.name}'" for item in _active_keymaps]), width=max_len)
+    output_str = f"{indent}{wrapped_items[0]}"
+    for item in wrapped_items[1:]:
```
I feel this is misguided... if something wraps weirdly you'd want that to *stand out*, not be hidden.  So I don't think any of this extra complexity is needed...

RedBearAK replied:
> I feel this is misguided... if something wraps weirdly you'd want that to _stand out_, not be hidden. So I don't think any of this extra complexity is needed...

I actually have no idea what you mean by this. Wraps weirdly? What’s being hidden?

RedBearAK replied:
If there's no wrapping algorithm at all, the terminal will even cut words apart, after the output reaches the right edge of the window. The wrapper at least cuts at word boundaries. 

Not sure what you want done here, or what you're objecting to.

RedBearAK replied:
I can add some ellipses characters to connect the wrapped items.  

```sh
(DD) ACTIVE KEYMAPS:
     'User hardware keys', 'Wordwise - not vscode', 'GenTerms overrides:
      … Ubuntu/Fedora', 'GenTerms overrides: GNOME', 'General Terminals', 'GenGUI
      … overrides: not Chromebook', 'GenGUI overrides: Fedora', 'GenGUI overrides:
      … GNOME', 'General GUI'
```

joshgoebel replied:
That's. better idea.

RedBearAK replied:
OK, it's in there. Anything else?

joshgoebel commented on src/keyszer/transform.py:
```diff
@@ -216,6 +216,31 @@ def dump_diagnostics():
+    import textwrap
+
+    debug("")
+    debug(f"WM_CLS: '{ctx.wm_class}' | WM_NME: '{ctx.wm_name}'")
+    debug(f"DVN: '{ctx.device_name}' | CLK: '{ctx.capslock_on}' | NLK: '{ctx.numlock_on}'")
```
```suggestion
    debug(f"WM_CLASS: '{ctx.wm_class}' | WM_NAME: '{ctx.wm_name}'")
    debug(f"DEVICE: '{ctx.device_name}' | CAPS_LOCK: '{ctx.capslock_on}' | NUM_LOCK: '{ctx.numlock_on}'")
```

Not worth the trouble vs just making it readable.

RedBearAK replied:
> Not worth the trouble vs just making it readable. 

That’s funny. I was literally just pondering doing that.

RedBearAK commented:
@joshgoebel 

Thanks, again! 🎉
